### ISSUE TITLE
API/UI: redraw "put" format is inefficient? (also: clarify docs)

### ISSUE THREAD

**Issue by realh** [neovim/neovim#6459]
- `nvim --version`:
NVIM 0.1.7
Build type: None
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-pN7CDA/neovim-0.1.7=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1  -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim-pN7CDA/neovim-0.1.7/build/config -I/build/neovim-pN7CDA/neovim-0.1.7/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim-pN7CDA/neovim-0.1.7/build/src/nvim/auto -I/build/neovim-pN7CDA/neovim-0.1.7/build/include
Compiled by pkg-vim-maintainers@lists.alioth.debian.org

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

- Vim (version: ) behaves differently? N/A
- Operating system/version: Linux lenny 4.9.0-2-amd64 #1 SMP Debian 4.9.13-1 (2017-02-27) x86_64 GNU/Linux
- Terminal name/version: [gnvim](https://github.com/realh/gnvim) remote UI
- `$TERM`: N/A

### Actual behaviour
The redraw put commands are packed as ["put", [string], [string], ...] where each argument is an array with a single string member containing one UTF-8 character.
### Expected behaviour
The documentation says the format should be ["put", string]. It isn't clear whether that implies a separate put command for every character, but that shouldn't be necessary:- clients can quite easily decompose the UTF-8 into single characters if necessary, and having one string for multiple characters is more efficient for the RPC. It would also actually be easier and more efficient for gnvim, because it's based on GtkTextView with the GtkTextBuffer holding text and highlight tags representing the current nvim view. Its API only supports erase/insert operations, not replace, so I reckon batching the characters into strings should be more efficient for it, and I'm currently doing that batching myself.

### Steps to reproduce using `nvim -u NORC`
N/A

**Comment by justinmk** [comment]
> having one string for multiple characters is more efficient for the RPC

Did you measure it?

**Comment by realh** [comment]
>> having one string for multiple characters is more efficient for the RPC
> Did you measure it?

No, but it's obvious. With a single string per span/line you only need one byte per character (plus a few extra in case of non-ASCII multibyte chars) in the msgpack plus one string marker code and string length per span/line. With the current system every single character additionally needs an array marker code + length and a string marker code + length. It must be using some 2-4x the bandwidth.

**Comment by justinmk** [comment]
It's not obvious that that will perceptibly affect performance,  even in performance-sensitive use cases. So it needs to be measured.

@lunixbochs did some profiling for actualvim, which has tight constraints, and this wasn't found to be a bottleneck, IIRC.

Also, IIRC there were valid reasons for sending nvim's utf8 representation. cc @tarruda

**Comment by bfredl** [comment]
> Also, IIRC there were valid reasons for sending nvim's utf8 representation.

Wasn't it something about double-width and ambiwith chars?

> needs an array marker code + length and a string marker code + length. 

msgpack has single-byte headers for really short strings and arrays, so it is not as bad as it first seems.

**Comment by lunixbochs** [comment]
ActualVim does not block on `put`, and grabs buffer contents separately, so the profiling data doesn't have any bearing on `put` performance.

**Comment by tarruda** [comment]
The format was created to be simple to parse using a msgpack library, not space efficient. Even so, it should be more than fast enough.

**Comment by tarruda** [comment]
> clients can quite easily decompose the UTF-8 into single characters if necessary, and having one string for multiple characters is more efficient for the RPC

Maybe so but then they would have to compute the cell widths. With the current format it is not necessary because cell offsets are handed over to the client

**Comment by realh** [comment]
>> clients can quite easily decompose the UTF-8 into single characters if necessary, and having one string for multiple characters is more efficient for the RPC
> Maybe so but then they would have to compute the cell widths. With the current format it is not necessary because cell offsets are handed over to the client

It's quite easy to compute the cell widths by scanning the string for multibyte characters, but I accept that for most implementations it's easier to have it already split into characters. gnvim has a bottleneck when scrolling, which I've posted about on reddit. Being able to process 'put' more efficiently might help, but probably not enough TBH.

**Comment by ZyX-I** [comment]
@realh It is by no means easy. How many cells a character occupies from the Neovim POV depends on the &ambiwidth setting, version of the unicode standard Neovim is using and the character itself. Given that version of the unicode standard will for sure be different on the client on some occasions it is better not offload that to the client. Also makes some bugs less probable.

**Comment by realh** [comment]
I didn't realise some characters were supposed to take up more than one cell. But I don't see how packing each character into its own array helps with that anyway, unless arrays containing wide characters have an extra element to flag them.

**Comment by ZyX-I** [comment]
@realh I tried to determine how it handles this situation experimentally, by adding a simple test

```patch
diff --git a/test/functional/ui/input_spec.lua b/test/functional/ui/input_spec.lua
index cec1925..d10d05d 100644
--- a/test/functional/ui/input_spec.lua
+++ b/test/functional/ui/input_spec.lua
@@ -117,6 +117,12 @@ describe('feeding large chunks of input with <Paste>', function()
       -- INSERT --                       20000,11      Bot |
     ]])
   end)
+
+  it("doublewidth", function()
+    feed('iＡＢ́́́')
+    screen:sleep(100)
+    screen:snapshot_util()
+  end)
 end)
 
 describe('input utf sequences that contain CSI/K_SPECIAL', function()
```

(note: this is fullwidth A, fullwidth B and two following identical combining characters).

That crashed! ASAN failure:

```
==================== File /home/zyx/a.a/Proj/c/neovim/asan.27341 ====================
= /home/zyx/a.a/Proj/c/neovim/src/nvim/api/ui.c:355:35: runtime error: null pointer passed as argument 1, which is declared to never be null
= /home/zyx/a.a/Proj/c/neovim/build/include/memory.h.generated.h:14:114: note: nonnull attribute specified here
= SUMMARY: AddressSanitizer: undefined-behavior /home/zyx/a.a/Proj/c/neovim/src/nvim/api/ui.c:355:35 in
=====================================================================================
```

Built with `CC=clang CFLAGS=" -O3 -g -DEXITFREE " cmake .. -DMIN_LOG_LEVEL=0 -DCMAKE_BUILD_TYPE=Release -DBUSTED_OUTPUT_TYPE=gtest -DCMAKE_INSTALL_PREFIX=$PWD/root -DCLANG_ASAN_UBSAN=ON -DPREFER_LUAJIT=false` (note: release build), on commit b5a0cc345f089afdce9e2bf39b125fcd8028f472 (current #6439 HEAD).

**Comment by ZyX-I** [comment]
Note: ui/input_spec.lua is not the right place for the test. I just wanted some random UI test file with screen already configured.

**Comment by ZyX-I** [comment]
Note 2: “experimentally” means “by checking the log file” (note zero minimal log level), snapshot_util is there only because I wanted to check that screen test infrastructure handles that nicely.

**Comment by bfredl** [comment]
Using `redraw_debug()` (and non-ASAN build) gives (relevant part)

```
{ "highlight_set", { {} } }
{ "cursor_goto", { 0, 0 } }
{ "put", { "Ａ" }, { "" }, { "Ｂ́́" }, { "" } }

screen:expect([[
  ＡＢ́́^                                    |
  {1:~                                       }|
  {1:~                                       }|
  {1:~                                       }|
  {1:~                                       }|
  {1:~                                       }|
  {1:~                                       }|
  {7:-- INSERT --}                            |
]])

```

(it might not look good in terminal or browser, but inspecting the UTF-8 it looks correct)

**Comment by bfredl** [comment]
(also quick note: `snapshot_util()` already does `screen:sleep()` (without it it would be pretty useless) so doing it manually is not necessary. )

**Comment by tarruda** [comment]
> gnvim has a bottleneck when scrolling, which I've posted about on reddit. Being able to process 'put' more efficiently might help, but probably not enough TBH.

For most scrolling cases, `put` should not affect performance as you think, and it should only be used for filling the "emptied areas"  which are consequence of moving a block of cells up or down. You should have a look at how the python gui is implemented for an idea.

**Comment by realh** [comment]
> For most scrolling cases, put should not affect performance as you think, and it should only be used for filling the "emptied areas" which are consequence of moving a block of cells up or down. You should have a look at how the python gui is implemented for an idea.
I think you're right, but GtkTextBuffer's API doesn't directly support moving or replacing text, so I have to erase and insert, which is inefficient for both cases. Overall it "feels" more sluggish than classic gvim. GtkTextBuffer does have a method to copy from one part of a buffer to another, which I use for scrolling, but again I have to add an erase to turn it into a move. I can probably make a big improvement by having a special case for when the scroll region is the entire width of the viewport, so I'll try that next. It won't help for vertically split windows though, and I use those a lot. Ideally I'd like to use GUI scrollbars like classic vim GUIs, but I think there are too many difficulties in the way of supporting multiple scrollbars for multiple split windows.

**Comment by realh** [comment]
> Using redraw_debug() (and non-ASAN build) gives (relevant part) ...

That definitely looks screwy, both in Thunderbird and Chrome; the former displays what looks like a capital B after the A, Chrome displays a box, so I guess the character is missing from the relevant font I use in Chrome. But in the line containing put, I see an opening quote, then the B or box, then what looks like it's tried to put an acute accent on the closing quote. Under screen:expect I see what looks like AB^ where the ^ is actually something that looks a bit like a Greek lambda in superscript.

**Comment by tarruda** [comment]
> I think you're right, but GtkTextBuffer's API doesn't directly support moving or replacing text,

There's the problem: `GtkTextBuffer` is not designed for this use case. I suggest you to look at how python GUI is implemented: https://github.com/neovim/python-gui. It's also a gtk application but uses python bindings. Cairo/pango are used to draw text, and cairo surface coordinate transformations are used when scrolling.

If you are implementing this in C, why not make it a reusable Gtk Widget? That way other GTK applications could embed neovim very easily :)

**Comment by bfredl** [comment]
Yes yes it looks pretty wrong in many terminals, but if you inspect it with hexdump ones sees that nvim and nvim test infrastructure has treated the data correctly to unicode rules. nvim cannot help that the terminal later displays it wrong...

**Comment by realh** [comment]
@tarruda Yes, I think I would have to use an approach like python-gui's to make it viable. But, if I've interpreted the reddit thread correctly, I would still have a problem with scrolling unless I do a lot of extra work to combine the redraws. Overall I don't think I could ever make something I could be truly satisifed with, because nvim just doesn't support the use of scrollbars :-(.

I did derive directly from Gtk classes, and would make a new type of widget for doing the cairo rendering myself, but I used gtkmm (C++). C++ is more intellectually "interesting" to me than all that boilerplate and manual vtable setup that goes with writing GObjects in C.
